### PR TITLE
Sub-event aware genJet and parton matching for HI miniAOD

### PR DESCRIPTION
**Pull request by mandrenguyen**
#### PR description:

For PbPb generators such as Hydjet we cluster the genJets for each nucleon-nucleon interaction separately.   This is done by SubEventGenJetProducer.cc, which gives collections like ak4HiGenJets, already in use in heavy-ion workflows.
In order to do a sensible genJet-recoJet association in PAT, one needs additional procedures that depends on what type of MC is run (ditto for parton association).   
For "minimum-bias" generation (a single PbPb event), one needs a cleaning procedure to avoid matching energetic recoJets to incidental overlapping low pT genJets.
For "embedded mode" where we overlay a pp generator (e.g., pythia) onto a PbPb event, we need to use the gen info from the signal event only, despite the fact that we store both the signal and background event in genParticles. 
   
The code modifications consist of: 
- The genJetSubEvent modifier was added, and activated for all MB Hydjet RelVal worklows that use the pp reconstruction (the older --scenario HeavyIons workflows are not able to run miniAOD).  
- Producers were added to filter out the signal gen information from genParticles.  These are active if the above modifier is not called.  In a pp event without event overlay, the code will return the usual gen info.  
- A counter was added to PATGenJetSlimmer.cc such that it no longer assumes pT-ordered genJets, which is not the case for the sub-event genJet producer 
- SubEventGenJetProducer no longer produces a collection of BasicJets, which were not used for anything, and were causing trouble for modules that don't get by type. 
- slimmedGenJets now takes as input the heavy-ion style genJets
- slimmedGenJetsFlavourInfos was disactivated, since it doesn't work for sub-event genJets.  It's not consumed by any downstream module in heavy-ion wfs. 

#### PR validation:

Tested with 158, 159,  158.01 and 301.   

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: Configuration/Eras/python/Modifier_genJetSubEvent_cff.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import FWCore.ParameterSet.Config as cms
+
+genJetSubEvent =  cms.Modifier()
```
Comment (slava77) [inline Configuration/Eras/python/Modifier_genJetSubEvent_cff.py:3]: another case for ProcessModifier

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  int jetsize = genjets->size();
+
+  vector<int> selection;
+  selection.reserve(jetsize);
+  for (int ijet = 0; ijet < jetsize; ++ijet) {
+    selection.push_back(-1);
+  }
+
+  vector<int> selectedIndices;
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:106]: Local vector never used

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  vector<int> removedIndices;
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:107]: Local vector is filled at L121, but never used

File: RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+#include "DataFormats/Math/interface/LorentzVector.h"
+
+using namespace std;
+using namespace edm;
+
+//
+// class decleration
+//
+
+class HiSignalParticleProducer : public edm::EDProducer {
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc:47]: legacy producer...

File: RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+public:
+  explicit HiSignalParticleProducer(const edm::ParameterSet&);
+  ~HiSignalParticleProducer() override;
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc:50]: `= default`

File: RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  int genParticleSize = genParticles->size();
+
+  vector<int> selection;
+  selection.reserve(genParticleSize);
+  for (int igenParticle = 0; igenParticle < genParticleSize; ++igenParticle) {
+    selection.push_back(-1);
+  }
+
+  vector<int> selectedIndices;
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc:105]: As in the previous producer, these two local vectors have no usage here

File: RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  using namespace reco;
+
+  auto signalGenParticles = std::make_unique<GenParticleCollection>();
+
+  edm::Handle<edm::View<GenParticle> > genParticles;
+  iEvent.getByToken(genParticleSrc_, genParticles);
+
+  int genParticleSize = genParticles->size();
+
+  vector<int> selection;
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalParticleProducer.cc:99]: As in the previous producer, this `selection` vector has no usage here

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  using namespace reco;
+
+  auto jets = std::make_unique<GenJetCollection>();
+
+  edm::Handle<edm::View<GenJet> > genjets;
+  iEvent.getByToken(jetSrc_, genjets);
+
+  int jetsize = genjets->size();
+
+  vector<int> selection;
```
[the remark this answers is not in the document]
Comment (perrotta) [reply RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:100]: In any case this `selection` local vector doesn't seem to be used anywhere, besides being filled and updates

File: RecoJets/JetProducers/plugins/SubEventGenJetProducer.cc [hunk 9]
```diff
@@ -38,7 +38,7 @@ namespace {
 SubEventGenJetProducer::SubEventGenJetProducer(edm::ParameterSet const& conf) : VirtualJetProducer(conf) {
   //   mapSrc_ = conf.getParameter<edm::InputTag>( "srcMap");
```
Comment (perrotta) [inline RecoJets/JetProducers/plugins/SubEventGenJetProducer.cc:39]: remove this commented out code if not needed

File: RecoJets/JetProducers/plugins/SubEventGenJetProducer.cc [hunk 10]
```diff
@@ -38,7 +38,7 @@ namespace {
...
   ignoreHydro_ = conf.getUntrackedParameter<bool>("ignoreHydro", true);
-  produces<reco::BasicJetCollection>();
+
```
Comment (perrotta) [inline RecoJets/JetProducers/plugins/SubEventGenJetProducer.cc:41]: What is this going to `produces` now?

File: RecoJets/JetProducers/plugins/SubEventGenJetProducer.h [hunk 11]
```diff
@@ -10,6 +10,7 @@
 
  ************************************************************/
 
+#include <memory>
```
Comment (perrotta) [inline RecoJets/JetProducers/plugins/SubEventGenJetProducer.h:13]: As such `<memory>` is included twice, in the header and in the implementation: please remove from the implementation file

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/JetReco/interface/GenJetCollection.h"
+#include "DataFormats/GeometryVector/interface/VectorUtil.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+
+#include "DataFormats/Math/interface/Point3D.h"
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:38]: These two includes do not seem to be used here: please remove if so
Perhaps, also some of the includes above are not needed: please check

**Review by perrotta — COMMENTED** [review]
The two producers `HiSignalGenJetProducer` and `HiSignalParticleProducer` are pretty identical, modulo the input and output collections: please template them from a single  `HiSignalProducer` (if you like the name...)

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+using namespace edm;
+
+//
+// class decleration
+//
+
+class HiSignalGenJetProducer : public edm::EDProducer {
+public:
+  explicit HiSignalGenJetProducer(const edm::ParameterSet&);
+  ~HiSignalGenJetProducer() override;
```
[the remark this answers is not in the document]
Comment (makortel) [reply RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:51]: The explicit destructor could also be removed, compiler will generate the `default` destructor implicitly in that case.

**Comment by mandrenguyen** [comment]
> The two producers `HiSignalGenJetProducer` and `HiSignalParticleProducer` are pretty identical, modulo the input and output collections: please template them from a single `HiSignalProducer` (if you like the name...)

The functionality in is not exactly identical.  For the genParticles, we loop over them and store the ones from the sub-event with collision ID =0.  For the genJets, we loop over them and for each one check whether the 1st constituent has collision ID =0. 
Since they are doing something slightly different, it's not obvious to me how to do this with a single templated producer.  Perhaps that's my own shortcoming though.

File: RecoJets/JetProducers/plugins/SubEventGenJetProducer.cc (hunk 10 above)
Comment (mandrenguyen) [reply RecoJets/JetProducers/plugins/SubEventGenJetProducer.cc:41]: The code is producing  a collection of type
 vector<reco::GenJet>
If I add the line:
produces<reco::GenJetCollection>();
I get an exception complaining about a 
"Duplicate Product Identifier"

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+  using namespace reco;
+
+  auto jets = std::make_unique<GenJetCollection>();
+
+  edm::Handle<edm::View<GenJet> > genjets;
+  iEvent.getByToken(jetSrc_, genjets);
+
+  int jetsize = genjets->size();
+
+  for (int ijet = 0; ijet < jetsize; ++ijet) {
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:93]: Using a range based loop can allow simpler and more readable syntax

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+// constructors and destructor
+//
+
+HiSignalGenJetProducer::HiSignalGenJetProducer(const edm::ParameterSet& iConfig)
+    : jetSrc_(consumes<edm::View<reco::GenJet> >(iConfig.getParameter<edm::InputTag>("src"))) {
+  std::string alias = (iConfig.getParameter<InputTag>("src")).label();
+  produces<reco::GenJetCollection>().setBranchAlias(alias);
+}
+
+//
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:76]: Also these commented out lines can be removed to clean up

**Review by perrotta — COMMENTED** [review]
Ok for keeping separated the two plugins.
Please find here another set of comments, mostly style related. What commented for `HiSignalGenJetProducer` can be repeated identically for `HiSignalParticleProducer`

**Comment by perrotta** [comment]
please test

**Comment by mandrenguyen** [comment]
For info, the HiSignal* code is not active in the reMiniAOD wfs (140.5611 and 158.01), as the gen cleaning is used instead via the process modifer.  I  tested wf 300, where this code is used.  I also removed the process modifier from the driver to test this code with 158.01, just to make sure it runs in reMiniAOD.

**Comment by kpedro88** [comment]
+upgrade

File: Configuration/PyReleaseValidation/python/relval_steps.py [hunk 16]
```diff
@@ -2484,13 +2490,26 @@ def gen2021HiMix(fragment,howMuch):
                                                         '--datatier':'GEN-SIM-RECO,MINIAODSIM,DQMIO,ALCARECO',
                                                         '--eventcontent':'RECOSIM,MINIAODSIM,DQM,ALCARECO',
                                                         },step3Up2015Defaults])
-steps['RECOHI2018PPRECOml']=merge([concurrentLumis,steps['RECOHI2018PPRECO']])
+steps['RECOHI2018PPRECOMB']=merge([hiDefaults2018_ppReco,{'-s':'RAW2DIGI,L1Reco,RECO,ALCA:SiStripCalZeroBias+SiPixelCalZeroBias,EI,PAT,VALIDATION:@standardValidationNoHLT+@miniAODValidation,DQM:@standardDQMFakeHLT+@miniAODDQM',
+                                                        '--datatier':'GEN-SIM-RECO,MINIAODSIM,DQMIO,ALCARECO',
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/relval_steps.py:2494]: can you check indentation among these lines? not sure if python will complain but better fix

File: Configuration/PyReleaseValidation/python/relval_steps.py [hunk 17]
```diff
@@ -2484,13 +2490,26 @@ def gen2021HiMix(fragment,howMuch):
...
 steps['REMINIAODHI2018PPRECO']=merge([{'-s':'PAT,VALIDATION:@miniAODValidation,DQM:@miniAODDQM',
                                                              '--datatier':'MINIAODSIM,DQMIO',
                                                              '--eventcontent':'MINIAODSIM,DQM',
                                                              '-n':100,
                                                          },hiDefaults2018_ppReco,step3Up2015Defaults])
+steps['REMINIAODHI2018PPRECOMB']=merge([{'-s':'PAT,VALIDATION:@miniAODValidation,DQM:@miniAODDQM',
+                                                             '--datatier':'MINIAODSIM,DQMIO',
+                                                             '--eventcontent':'MINIAODSIM,DQM',
+                                                             '-n':100,
+                                         '--era':'Run2_2018_pp_on_AA',
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/relval_steps.py:2510]: also here

**Review by chayanit — CHANGES_REQUESTED** [review]

**Comment by chayanit** [comment]
please test

File: RecoHI/HiJetAlgos/python/HiSignalGenJetProducer_cfi.py [hunk 18]
```diff
@@ -0,0 +1,7 @@
+import FWCore.ParameterSet.Config as cms
+
+
+hiSignalGenJets = cms.EDProducer('HiSignalGenJetProducer',
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/python/HiSignalGenJetProducer_cfi.py:4]: File name should correspond to the module name: this should then be `hiSignalGenJets_cfi.py`.

In any case, if you don't plan to modify these file any further (they are just two parameters which I don't expect will change even for the coming runs) you can stick on the configuration automatically created by the `fillDescriptions` method.

On the contrary, if you still plan to update these configs, you can keep this explicit python config file, but rather clone the one created by `fillDescriptions`, e.g.
```PYTHON
import FWCore.ParameterSet.Config as cms
from RecoHI.HiJetAlgos.hiSignalGenJetProducer_cfi import hiSignalGenJetProducer
hiSignalGenJets = hiSignalGenJetProducer.clone()
```
(and this file could become a `*_cff` then).

If needed in the future you will modify the parameters in the cloned object if needed

File: RecoHI/HiJetAlgos/python/HiSignalParticleProducer_cfi.py [hunk 19]
```diff
@@ -0,0 +1,7 @@
+import FWCore.ParameterSet.Config as cms
+
+
+hiSignalGenParticles = cms.EDProducer('HiSignalParticleProducer',
```
Comment (perrotta) [inline RecoHI/HiJetAlgos/python/HiSignalParticleProducer_cfi.py:4]: Same comment as for the sibling config

File: RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  iEvent.put(std::move(jets));
+}
+
+void HiSignalGenJetProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.setComment("Selects genJets from collision id = 0");
+  desc.add<edm::InputTag>("src", edm::InputTag("akHiGenJets"));
+  descriptions.add("HiSignalGenJetProducer", desc);
```
[the remark this answers is not in the document]
Comment (slava77) [reply RecoHI/HiJetAlgos/plugins/HiSignalGenJetProducer.cc:59]: `.addWithDefaultLabel(desc);` is actually preferred if there is no reason to have a special name

**Comment by chayanit** [comment]
please test

**Comment by chayanit** [comment]
+1

**Comment by perrotta** [comment]
Checked wf 158 with and without the `genJetSubEvent` modifier, and everything behaves as expected: with that modifier the current (i.e. as in the baseline) `slimmedGenJets` in miniAOD are reproduced, while when it gets removed `hiSignalGenJets` are the source for those `slimmedGenJets`, as they should.

The net result is that when the `hiSignalGenJets` are the source for the `slimmedGenJets`, those `slimmedGenJets` have a larger set of partons from which they can be built, and therefore the `slimmedGenJets` size increases in miniAOD, as expected:
```
 Checking process _ step3_MB_inMINIAODSIM.root and step3_inMINIAODSIM.root with useUnpacked=no (if above 1 or 1%):
Compare packed values
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
  32277.0 ->     39497.0       7220     20.1   1.05     recoGenJets_slimmedGenJets__RECO.
-------------------------------------------------------------
   684470 ->      691690       7220             1.1     ALL BRANCHES
```

**Comment by perrotta** [comment]
+1
- Updates are consistent with the expectations described in the PR descriptions
- No differences in reco/miniAOD outputs in the workflows tested automatically, while changes are there for the HI workflows in  "embedded mode". The difference being only in the genJet collection in miniAOD, it is then probably not necessary to have one of those workflows included in the automatic tests: but I'd let the HI responsible think about it

**Comment by kpedro88** [comment]
+upgrade

**Comment by silviodonato** [comment]
+operations

**Comment by silviodonato** [comment]
merge

**Comment by santocch** [comment]
+1

**Pull request merged**